File: plugin/src/main/java/ru/arc/AuctionMessager.java
package ru.arc;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import lombok.RequiredArgsConstructor;

import java.lang.reflect.Type;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@RequiredArgsConstructor
public class AuctionMessager implements ChannelListener {
    public final String channelPartial;
    public final String channelAll;
    public final DiscordBot discordBot;

    Map<UUID, AuctionItemDto> map = new ConcurrentHashMap<>();

    @Override
    public void consume(String channel, String message, String originServer) {
        if (channel.equals(channelAll)) {
            map.clear();
        }

        Gson gson = new Gson();
        Type listType = new TypeToken<List<AuctionItemDto>>() {
        }.getType();
        List<AuctionItemDto> auctionItemDtos = gson.fromJson(message, listType);

        for (AuctionItemDto auctionItemDto : auctionItemDtos) {
            try {
                if (auctionItemDto.exist) {
                    map.put(UUID.fromString(auctionItemDto.uuid), auctionItemDto);
                } else {
                    map.remove(UUID.fromString(auctionItemDto.uuid));
                }
            } catch (Exception e) {
                System.out.println("Error: " + auctionItemDto);
                e.printStackTrace();
            }
        }

        List<AuctionItemDto> dtos = map.values().stream()
                .sorted(Comparator.comparingInt(dto -> dto.priority))
                .toList();

        if (discordBot != null) discordBot.updateAuctionItems(dtos);
        else {
            System.out.println("Discord bot is null!");
        }
    }
}

File: src/main/kotlin/ru/arc/auction/AuctionMessager.kt
package ru.arc.auction

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.slf4j.LoggerFactory
import ru.arc.discord.DiscordBot
import ru.arc.velocity.Velocity
import ru.arc.redis.ChannelListener
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class AuctionMessager(
    @JvmField val channelPartial: String,
    @JvmField val channelAll: String,
    @JvmField val channelSales: String = "arc.auction_sale_events",
    private val discordBotProvider: () -> DiscordBot? = { Velocity.discordBot },
) : ChannelListener {

    private val log = LoggerFactory.getLogger(AuctionMessager::class.java)
    private val gson = Gson()

    @JvmField
    val map: MutableMap<UUID, AuctionItemDto> = ConcurrentHashMap()

    private val snapshotsByOrigin = mutableMapOf<String, MutableMap<UUID, AuctionItemDto>>()

    @Synchronized
    override fun consume(channel: String, message: String, originServer: String) {
        if (channel == channelSales) {
            consumeSale(message, originServer)
            return
        }

        val listType = object : TypeToken<List<AuctionItemDto>>() {}.type
        val auctionItemDtos =
            runCatching { gson.fromJson<List<AuctionItemDto>>(message, listType) }
                .getOrElse { error ->
                    log.warn("Ignoring malformed auction update from {}", originServer, error)
                    return
                } ?: emptyList()

        val origin = originServer.ifBlank { UNKNOWN_ORIGIN }
        if (channel == channelAll) {
            val replacement = mutableMapOf<UUID, AuctionItemDto>()
            applyUpdate(replacement, auctionItemDtos, origin)
            snapshotsByOrigin[origin] = replacement
        } else {
            val snapshot = snapshotsByOrigin.getOrPut(origin) { mutableMapOf() }
            applyUpdate(snapshot, auctionItemDtos, origin)
        }

        map.clear()
        snapshotsByOrigin.toSortedMap().values.forEach { snapshot ->
            snapshot.forEach { (id, item) -> map[id] = item }
        }

        val dtos = map.values.sortedBy { it.priority }
        discordBotProvider()?.updateAuctionItems(dtos)
            ?: log.debug("Skipping auction Discord update because the bot is not ready")
    }

    private fun applyUpdate(
        snapshot: MutableMap<UUID, AuctionItemDto>,
        updates: List<AuctionItemDto>,
        originServer: String,
    ) {
        for (item in updates) {
            try {
                val id = UUID.fromString(item.uuid)
                if (item.exist) {
                    snapshot[id] = item
                } else {
                    snapshot.remove(id)
                }
            } catch (e: Exception) {
                log.warn("Ignoring invalid auction item from {}: {}", originServer, item, e)
            }
        }
    }

    private fun consumeSale(
        message: String,
        originServer: String,
    ) {
        val event =
            runCatching { gson.fromJson(message, AuctionSaleEventDto::class.java) }
                .getOrElse { error ->
                    log.warn("Ignoring malformed auction sale event from {}", originServer, error)
                    return
                } ?: return
        if (!validSale(event)) {
            log.warn("Ignoring invalid auction sale event from {}", originServer)
            return
        }
        discordBotProvider()?.notifyAuctionSold(event)
    }

    companion object {
        internal fun validSale(event: AuctionSaleEventDto): Boolean =
            event.listingId.length in 1..64 &&
                PLAYER_NAME.matches(event.sellerName) &&
                PLAYER_NAME.matches(event.buyerName) &&
                event.itemDisplay.isNotBlank() && event.itemDisplay.length <= 200 &&
                event.amount in 1..1_000_000 && event.price.isNotBlank() &&
                event.price.length <= 100 && event.occurredAt >= 0

        private val PLAYER_NAME = Regex("[A-Za-z0-9_]{1,16}")
        private const val UNKNOWN_ORIGIN = "unknown"
    }
}
